rfd/0078-role-variables.md

```diff
@@ -0,0 +1,358 @@
+          out:
+            - "bar"
+            - "{{vars.var_b}}"
+```
+
+### Trusted clusters
+
+Variables defined in the root cluster will not be usable in leaf clusters.
+It is expected that the variable will be redefined in the leaf cluster.
+
+It's common for deployments to copy the exact same role definitions to all
+clusters, so this seems like the most logical choice.
+As long as the role which defines the variable is present in both clusters and
+it is mapped to its leaf equivalent, you can seamlessly use the variable in both
+clusters.
+
+## Examples
+
+### Defined allow labels for each group
+
+If the Teleport admin cannot create custom traits in their IDP, then it is
+extremely cumbersome and repetitive (borderline impossible) to define a custom
+mapping per group.
+
+Solution today:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: users
+spec:
+  allow:
+    node_labels:
+      env:
+        - regexp.replace(external.groups, "dev", "dev")
+        - regexp.replace(external.groups, "dev", "staging")
+        - regexp.replace(external.groups, "qa", "qa")
+        - regexp.replace(external.groups, "qa", "staging")
+    app_labels:
+      env:
+        - regexp.replace(external.groups, "dev", "dev")
+        - regexp.replace(external.groups, "dev", "staging")
+        - regexp.replace(external.groups, "qa", "qa")
+        - regexp.replace(external.groups, "qa", "staging")
+```
+
+With more roles and more complex rules, this would be even more difficult. The
+full list would have to be repeated for every value in every role. I don't think
+anyone even knows you can do this, and I hope no-one actually does, because it's
+ridiculous.
+
+With role variables it becomes trivial:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: users
+spec:
+  vars:
+    - name: allow_envs
+      values:
+        - input: "{{external.groups}}"
+          match: "^(qa|devs)$"
+          out:
+            - "$1"
+            - "staging"
+  allow:
+    node_labels:
+      env: "{{vars.allow_envs}}"
+    app_labels:
+      env: "{{vars.allow_envs}}"
+```
+
+The same variable can easily be defined once, and used many times within this
+role.
+
+## Alternatives
+
+### Trait transformations
+
+Instead of defining variables in the role spec, we could create a new `trait`
+resource which can define new traits which will be computed during login and
+embedded in the user's cert.
+
+These traits would be reusable across roles and trusted clusters.
+
+The syntax could be similar to the proposed role variable syntax:
+
+```yaml
+kind: trait
+version: v1
+metadata:
+  name: groups
+spec:
+  values:
+    # An input will expand to a list of N inputs if the trait is a list of N.
+    # Any input value which matches the `match` regex will contribute one
+    # output value, which may contain regex captures from that specific input.
+    #
+    # Here [devs, env-staging, env-prod] maps to [staging, prod].
+    - input: "{{external.groups}}"
```

Comment:
if input is always a variable, may be we don't need any brackets or other syntax?